Chapter-4-applying-tactical-domain-driven-design/README.adoc

```diff
@@ -40,29 +77,238 @@ IMPORTANT: It makes no sense to consider the Domain Model in typical CRUD module
+
+[[value-object]]
+==== Value Object
+
+A **Value Object** represents a business concept without a lifecycle. Unlike entities, value objects lack identity. They are immutable, serving as explicit types that describe specific aspects of the domain. 
+
+They enhance our domain model's expressiveness, prevent invalid object state and helps us avoid primitive obsession.
+
+"Primitive Obsession" is a code smell where we use primitive types to represent domain concepts. 
+
+image::Assets/value_object_sets.png[300,300,align=center]
+
+For e.g you can treat _SSN_ as _string_. _String_ allow put to field every character, but _SSN_ has specific format. Take a look at image above. You can see that _SSN_ is just only small subset of _string_ possible values.
+**Value object is precise and can validate format during object initialization.**
+
+==== Examples from different domains:
+- Iccid (International Circuit Card Identifier) in the telecommunications domain (this not just string but every character metters)
+- SSN (Social Security Number) in the financial domain
+- Money in the financial domain (amount and currency)
+- Address in the e-commerce domain
+- PhoneNumber in the telecommunication domain
+- Email in the e-commerce domain
+- Energy Indicator in the eco domain (amount and unit)
+
+==== Main characteristics of value objects:
+- They have no identity
+- They represent a bussiness concept
+- They can encapsulate bussiness logic like validation during object initialization
+- They can be used as a part of an entity or aggregate root
+- They are immutable
+- They are compared by their properties
+- They have equals and hashcode methods implemented
+- In one bounded context concept can be modeled as value object and in another as entity
+
+[source,csharp]
+----
+public sealed class Signature : ValueObject
+{
+    public string Value { get; init; }
+    public DateTimeOffset SignedAt { get; init; }
+
+    private Signature(string value)
+    {
+        if (string.IsNullOrWhiteSpace(value))
+            throw new ArgumentException("Signature cannot be empty");
+
+        Value = value;
+    }
+
+    public static Signature Create(string value) => new(value);
+
+    protected override IEnumerable<object> GetEqualityComponents()
+    {
+        yield return Value;
+    }
+}
+----
+
+[[aggregate-root]]
+==== Aggregate Root
+
+An **Aggregate Root** is a collection of related objects that are treated as a single unit. It consists of entities and value objects that are interconnected through a root entity. **Consider the aggregate root as a safeguard for invariants, ensuring the consistency and validity of the domain model.** Those invariants are esential bussiness rules, we can't model real business processes without them.
+
+Aggregate roots are also transactional boundaries. All changes to the aggregate should be done through the aggregate root. This ensures that the aggregate is always in a consistent state.
```

Comment:
Add a description of aggregate root as well. That the each aggregate has its root and this root is represented by an entity